File: backend/src/igab/api/v1/schemas/wishlist.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import Field, model_validator

from igab.api.v1.schemas.base import ApiModel, ClientDated
from igab.guide.wishlist import MAX_COOLING_DAYS, MAX_REVIEW_DAYS, MIN_REVIEW_DAYS

Money = Decimal


class FundingIn(ApiModel):
    """Where a wish's money lives.

    `own` makes a category of its own in the Wishlist group with a savings
    goal equal to the cost (and an optional date); `existing` points at any
    category; `none` leaves it for later.
    """

    mode: Literal["own", "existing", "none"] = "none"
    category_id: uuid.UUID | None = None
    want_by: date | None = None

    @model_validator(mode="after")
    def shape(self) -> "FundingIn":
        if self.mode == "existing" and self.category_id is None:
            raise ValueError("existing funding needs a category_id")
        return self


class WishCreate(ClientDated):
    """`client_today` is the day the wish is added: what `cooling_days`
    counts from, now and on every later edit that sets the period in days."""

    name: str = Field(min_length=1, max_length=200)
    cost: Money = Field(default=Decimal("0"), ge=0)
    url: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=2000)
    project_id: uuid.UUID | None = None
    priority: int | None = Field(default=None, ge=0)
    cooling_days: int | None = Field(default=None, ge=0, le=MAX_COOLING_DAYS)
    funding: FundingIn = Field(default_factory=FundingIn)


class WishUpdate(ClientDated):
    """`client_today` is the person's own date, and a status change stamps it.

    Without it `done_at`/`dropped_at` came from the server's clock, and the
    discipline report buckets "cooled off, then dropped" by comparing that
    date with `cooling_until` — so a drop on the last cooling evening west of
    UTC was filed as a wish that never cooled at all.
    """

    name: str | None = Field(default=None, min_length=1, max_length=200)
    cost: Money | None = Field(default=None, ge=0)
    url: str | None = Field(default=None, max_length=2000)
    notes: str | None = Field(default=None, max_length=2000)
    project_id: uuid.UUID | None = None
    priority: int | None = Field(default=None, ge=0)
    #: Pin as a top priority. Capped server-side (PRIORITY_LIMIT); a full
    #: spotlight refuses the pin rather than silently displacing one.
    is_priority: bool | None = None
    status: Literal["open", "done", "dropped"] | None = None
    cooling_until: date | None = None
    #: The cooling-off as a length: this many days after the wish was ADDED,
    #: not after today — so "14" means the same date whenever the edit is
    #: made, and one that is already behind us simply ends the cooling-off.
    #: The other spelling of `cooling_until`; sending both is refused.
    cooling_days: int | None = Field(default=None, ge=0, le=MAX_COOLING_DAYS)
    #: Any of the three modes. `own` on a wish that has no envelope of its
    #: own makes one (named for the wish, with a savings goal of its cost);
    #: on one that already has an envelope it is a no-op, because the budget
    #: page owns that category from the moment it exists.
    funding: FundingIn | None = None

    @model_validator(mode="after")
    def one_cooling_spelling(self) -> "WishUpdate":
        # Presence, not value: `cooling_until: null` (end it) beside a day
        # count is as contradictory as two dates.
        if {"cooling_days", "cooling_until"} <= self.model_fields_set:
            raise ValueError("Set the cooling-off as a date or as days, not both")
        return self


class FundingOut(ApiModel):
    mode: Literal["own", "existing", "none"]
    category_id: uuid.UUID | None
    category_name: str | None
    inherited: bool
    owns_envelope: bool
    target_date: date | None


class ReachOut(ApiModel):
    state: Literal["now", "months", "no_rate", "unlinked"]
    months: int | None
    date: date | None
    ahead_cost: Decimal
    progress: Decimal


class SettlementOut(ApiModel):
    """An ended wish whose own envelope is still standing.

    Served, not derived on the client: `available` is the budget page's
    figure, and whether a goal is still attached is the server's to know.
    Null on `WishOut` once the envelope has been settled or archived, so the
    prompt clears itself rather than needing a flag that could disagree.
    """

    category_id: uuid.UUID
    name: str
    available: Decimal
    has_goal: bool


class WishOut(ApiModel):
    id: uuid.UUID
    project_id: uuid.UUID | None
    name: str
    url: str | None
    notes: str | None
    cost: Decimal
    priority: int
    #: Required, not optional: a path that forgets to serialize it must raise,
    #: not quietly report an unpinned wish.
    is_priority: bool
    status: Literal["open", "done", "dropped"]
    funding: FundingOut
    cooling_until: date | None
    cooling: bool
    #: The day the wish was added, as its cooling-off in days counts it —
    #: served because the client cannot recover the person's date from
    #: `created_at`, an instant.
    added_on: date
    last_affirmed_at: datetime | None
    review_due: bool
    done_at: date | None
    dropped_at: date | None
    created_at: datetime
    reach: ReachOut | None
    #: Required, not optional: a path that forgets it would report an
    #: envelope left holding money as settled, which is the bug this field
    #: exists to end.
    settlement: SettlementOut | None


class ProjectSummaryOut(ApiModel):
    item_count: int
    open_count: int
    total_cost: Decimal
    affordable_now: int
    funded_by: date | None
    state: Literal["now", "months", "no_rate", "unlinked", "mixed", "complete", "empty"]
    complete: bool


class ProjectOut(ApiModel):
    id: uuid.UUID
    name: str
    category_id: uuid.UUID | None
    category_name: str | None
    notes: str | None
    sort_order: int
    summary: ProjectSummaryOut


class ProjectCreate(ApiModel):
    name: str = Field(min_length=1, max_length=120)
    category_id: uuid.UUID | None = None
    notes: str | None = Field(default=None, max_length=2000)


class ProjectUpdate(ApiModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    category_id: uuid.UUID | None = None
    notes: str | None = Field(default=None, max_length=2000)


class WishReorder(ApiModel):
    item_ids: list[uuid.UUID]


class ProjectReorder(ApiModel):
    project_ids: list[uuid.UUID]


class WishlistSettingsOut(ApiModel):
    cooling_days: int
    review_after_days: int


class WishlistSettingsUpdate(ApiModel):
    cooling_days: int | None = Field(default=None, ge=0, le=MAX_COOLING_DAYS)
    review_after_days: int | None = Field(default=None, ge=MIN_REVIEW_DAYS, le=MAX_REVIEW_DAYS)


class SettleRequest(ClientDated):
    """Where an ended wish's envelope money goes, and what becomes of the
    envelope. `destination_category_id` null means Ready to Assign — where
    the category-delete flow and the wishlist off-switch both send it."""

    destination_category_id: uuid.UUID | None = None
    #: Keep the (now empty, goal-less) envelope on the budget page instead of
    #: archiving it — for someone who wants to re-purpose it.
    keep_envelope: bool = False


class EnvelopeOut(ApiModel):
    category_id: uuid.UUID
    name: str
    available: Decimal


class DeleteWishResponse(ApiModel):
    """The envelope the wish owned, if any, so the client can offer to
    delete it too through the ordinary category-delete flow."""

    envelope: EnvelopeOut | None


class DrainMoveOut(ApiModel):
    move_id: uuid.UUID
    month: date
    date: datetime
    amount: Decimal
    from_category_id: uuid.UUID
    from_name: str
    to_category_id: uuid.UUID | None
    to_name: str
    affected: list["DrainAffectedOut"]


class DrainAffectedOut(ApiModel):
    item_id: uuid.UUID
    name: str
    months_further: Decimal | None


class DrainsOut(ApiModel):
    month: date
    total: Decimal
    moves: list[DrainMoveOut]


class StillWantedOut(ApiModel):
    count: int
    of: int
    #: The window, in months — served so the client's copy has one source.
    months: int


class WishlistResponse(ApiModel):
    enabled: bool
    items: list[WishOut]
    history: list[WishOut]
    projects: list[ProjectOut]
    still_wanted: StillWantedOut
    review_due_count: int
    settings: WishlistSettingsOut
    #: The pin cap, served so the client disables the action at the limit
    #: without spelling its own 3.
    priority_limit: int
    #: The longest cooling-off, in days — served for the same reason.
    max_cooling_days: int
    #: The review cadence's bounds, served for the same reason again: the
    #: settings form refuses what the server would refuse, in its own words.
    min_review_days: int
    max_review_days: int
    drains: DrainsOut | None

File: frontend/src/api/wishlist.ts
import toast from 'react-hot-toast'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient, apiErrorMessage } from './client'
import { invalidateAfterCategoryChange } from './invalidateAfterCategoryChange'
import { ROOT } from './queryKeys'

// The wishlist lives inside the budget: a wish's money is an envelope's
// money. Everything below is served — reach, rollups, cooling, review-due —
// and the client only sorts, filters and formats it. Money arrives as a JSON
// number, like every `ApiModel` response (backend `schemas/base.py`). These
// said `string` long after that stopped being true, and the edit form seeded
// its Cost input from one: Save called `.trim()` on a number and threw before
// any request was sent.

export type FundingMode = 'own' | 'existing' | 'none'
export type WishStatus = 'open' | 'done' | 'dropped'
export type ReachState = 'now' | 'months' | 'no_rate' | 'unlinked'

export interface WishFunding {
  mode: FundingMode
  category_id: string | null
  category_name: string | null
  /** The envelope came from the project rather than the wish itself. */
  inherited: boolean
  /** The wishlist created this envelope; deleting the wish offers to delete it. */
  owns_envelope: boolean
  target_date: string | null
}

export interface WishReach {
  state: ReachState
  months: number | null
  date: string | null
  /** What the wishes ahead in the same envelope still need. */
  ahead_cost: number
  /** 0–1, net of the wishes ahead. */
  progress: number
}

export interface Wish {
  id: string
  project_id: string | null
  name: string
  url: string | null
  notes: string | null
  cost: number
  priority: number
  /** Pinned as a top priority — an explicit choice, capped at the served
   *  `priority_limit`; distinct from `priority`, the funding-queue order. */
  is_priority: boolean
  status: WishStatus
  funding: WishFunding
  cooling_until: string | null
  cooling: boolean
  /** The day the wish was added, in the person's own date: what a cooling-off
   *  in days counts from. Served, because `created_at` is an instant and the
   *  day it falls on depends on a timezone the server recorded at creation. */
  added_on: string
  last_affirmed_at: string | null
  review_due: boolean
  done_at: string | null
  created_at: string
  reach: WishReach | null
}

export type ProjectState =
  'now' | 'months' | 'no_rate' | 'unlinked' | 'mixed' | 'complete' | 'empty'

export interface ProjectSummary {
  item_count: number
  open_count: number
  total_cost: number
  affordable_now: number
  funded_by: string | null
  state: ProjectState
  complete: boolean
}

export interface WishlistProject {
  id: string
  name: string
  category_id: string | null
  category_name: string | null
  notes: string | null
  sort_order: number
  summary: ProjectSummary
}

export interface DrainAffected {
  item_id: string
  name: string
  months_further: number | null
}

export interface DrainMove {
  move_id: string
  month: string
  date: string
  amount: number
  from_category_id: string
  from_name: string
  to_category_id: string | null
  to_name: string
  affected: DrainAffected[]
}

export interface Drains {
  month: string
  total: number
  moves: DrainMove[]
}

export interface WishlistSettings {
  cooling_days: number
  review_after_days: number
}

export interface Wishlist {
  enabled: boolean
  items: Wish[]
  history: Wish[]
  projects: WishlistProject[]
  still_wanted: { count: number; of: number; months: number }
  review_due_count: number
  settings: WishlistSettings
  /** The pin cap, served — the client never spells its own 3. */
  priority_limit: number
  /** The longest cooling-off in days, served for the same reason. */
  max_cooling_days: number
  drains: Drains | null
}

export interface FundingIn {
  mode: FundingMode
  category_id?: string | null
  want_by?: string | null
}

export interface WishCreate {
  name: string
  cost: number
  url?: string | null
  notes?: string | null
  project_id?: string | null
  priority?: number | null
  cooling_days?: number | null
  /** The browser's today: the day the wish is added, which a cooling-off in
   *  days counts from now and on every later edit. */
  client_today?: string
  funding: FundingIn
}

export interface WishUpdate {
  name?: string
  cost?: number
  url?: string | null
  notes?: string | null
  project_id?: string | null
  priority?: number
  is_priority?: boolean
  status?: WishStatus
  cooling_until?: string | null
  /** The cooling-off as days after the wish was added — the other spelling of
   *  `cooling_until`. The server resolves the date and refuses both at once. */
  cooling_days?: number
  funding?: FundingIn
}

export interface ProjectCreate {
  name: string
  category_id?: string | null
  notes?: string | null
}

export interface ProjectUpdate {
  name?: string
  category_id?: string | null
  notes?: string | null
}

/** The envelope a deleted wish owned, offered for deletion with it. */
export interface WishEnvelope {
  category_id: string
  name: string
  available: number
}

export interface DeleteWishResult {
  envelope: WishEnvelope | null
}

export function useWishlist(budgetId: string | null, enabled = true) {
  return useQuery({
    queryKey: [ROOT.wishlist, budgetId],
    queryFn: () => apiClient.get<Wishlist>(`/${budgetId}/wishlist`).then((r) => r.data),
    enabled: !!budgetId && enabled,
    staleTime: 30_000,
  })
}

function useWishlistMutation<TVars, TResult>(
  budgetId: string,
  fn: (vars: TVars) => Promise<TResult>,
  failure: string,
  quiet = false
) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: fn,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: [ROOT.wishlist, budgetId] })
      // An own envelope is a real category with a goal: the budget page and
      // every picker need to hear about it.
      invalidateAfterCategoryChange(qc, budgetId)
    },
    onError: quiet ? undefined : (e) => toast.error(apiErrorMessage(e, failure)),
  })
}

export function useCreateWish(budgetId: string) {
  // Quiet: the form shows the server's reason inline (a name clash, say).
  return useWishlistMutation<WishCreate, Wish>(
    budgetId,
    (body) => apiClient.post<Wish>(`/${budgetId}/wishlist`, body).then((r) => r.data),
    'Could not add the wish',
    true
  )
}

export function useUpdateWish(budgetId: string) {
  return useWishlistMutation<{ id: string } & WishUpdate, Wish>(
    budgetId,
    ({ id, ...body }) =>
      apiClient.patch<Wish>(`/${budgetId}/wishlist/${id}`, body).then((r) => r.data),
    'Could not save the wish'
  )
}

export function useDeleteWish(budgetId: string) {
  return useWishlistMutation<string, DeleteWishResult>(
    budgetId,
    (id) => apiClient.delete<DeleteWishResult>(`/${budgetId}/wishlist/${id}`).then((r) => r.data),
    'Could not delete the wish'
  )
}

export function useAffirmWish(budgetId: string) {
  return useWishlistMutation<string, void>(
    budgetId,
    (id) => apiClient.post(`/${budgetId}/wishlist/${id}/affirm`).then(() => undefined),
    'Could not save'
  )
}

export function useReorderWishes(budgetId: string) {
  return useWishlistMutation<string[], void>(
    budgetId,
    (item_ids) =>
      apiClient.post(`/${budgetId}/wishlist/reorder`, { item_ids }).then(() => undefined),
    'Could not reorder'
  )
}

export function useCreateProject(budgetId: string) {
  return useWishlistMutation<ProjectCreate, WishlistProject>(
    budgetId,
    (body) =>
      apiClient.post<WishlistProject>(`/${budgetId}/wishlist/projects`, body).then((r) => r.data),
    'Could not add the project',
    true
  )
}

export function useUpdateProject(budgetId: string) {
  return useWishlistMutation<{ id: string } & ProjectUpdate, WishlistProject>(
    budgetId,
    ({ id, ...body }) =>
      apiClient
        .patch<WishlistProject>(`/${budgetId}/wishlist/projects/${id}`, body)
        .then((r) => r.data),
    'Could not save the project'
  )
}

export function useDeleteProject(budgetId: string) {
  return useWishlistMutation<string, void>(
    budgetId,
    (id) => apiClient.delete(`/${budgetId}/wishlist/projects/${id}`).then(() => undefined),
    'Could not delete the project'
  )
}

export function useReorderProjects(budgetId: string) {
  return useWishlistMutation<string[], void>(
    budgetId,
    (project_ids) =>
      apiClient
        .post(`/${budgetId}/wishlist/projects/reorder`, { project_ids })
        .then(() => undefined),
    'Could not reorder'
  )
}

export function useSetWishlistSettings(budgetId: string) {
  return useWishlistMutation<Partial<WishlistSettings>, WishlistSettings>(
    budgetId,
    (body) =>
      apiClient.put<WishlistSettings>(`/${budgetId}/wishlist/settings`, body).then((r) => r.data),
    'Could not save settings'
  )
}
